src/backend/opencl/kernel/reduce_blocks_by_key_first.cl

```diff
@@ -0,0 +1,121 @@
+/*******************************************************
+ * Copyright (c) 2018, ArrayFire
+ * All rights reserved.
+ *
+ * This file is distributed under 3-clause BSD license.
+ * The complete license agreement can be obtained at:
+ * http://arrayfire.com/licenses/BSD-3-Clause
+ ********************************************************/
+
+void inplace_blockscan_inclusive_add(__local Tk *arr) {
```

Comment:
Is there some code you can share with our current implementation of scan. Perhaps reuse/combine it to reduce repetition.